src/java/com/unifina/data/StreamrBinaryMessage.java

```diff
@@ -2,52 +2,100 @@
 
 import java.io.UnsupportedEncodingException;
 import java.nio.ByteBuffer;
-import java.nio.charset.StandardCharsets;
+import java.nio.charset.Charset;
+import javax.xml.bind.DatatypeConverter;
 
 public class StreamrBinaryMessage {
 	private static final byte VERSION = 28; //0x1C
+	private static final byte VERSION_SIGNED = 29; //0x1D
+
+	public static final byte SIGNATURE_TYPE_NONE = 0;
+	public static final byte SIGNATURE_TYPE_ETH = 1;
 
 	public static final byte CONTENT_TYPE_STRING = 11; //0x0B
 	public static final byte CONTENT_TYPE_JSON = 27; //0x1B
 
+	private static final Charset utf8 = Charset.forName("UTF-8");
+
+	private final byte version;
 	private final String streamId;
 	private final int partition;
 	private final long timestamp;
 	private final byte contentType;
 	private final byte[] streamIdAsBytes;
 	private final byte[] content;
 	private final int ttl;
+	private final byte signatureType;
+	private final byte[] addressBytes;
+	private final byte[] signatureBytes;
 
 	public StreamrBinaryMessage(ByteBuffer bb) {
-		byte version = bb.get();
+		version = bb.get();
 
 		// If the message starts with a version byte, parse timestamp and contentType headers
-		if (version==28) {
+		if (version == VERSION || version == VERSION_SIGNED) {
 			timestamp = bb.getLong();
 			ttl = bb.getInt();
 			int streamIdLength = bb.get() & 0xFF; // unsigned byte
 			streamIdAsBytes = new byte[streamIdLength];
 			bb.get(streamIdAsBytes);
-			streamId = new String(streamIdAsBytes, StandardCharsets.UTF_8);
+			streamId = new String(streamIdAsBytes, utf8);
 			partition = bb.get() & 0xff; // unsigned byte
 			contentType = bb.get();
 			int contentLength = bb.getInt();
 			content = new byte[contentLength];
 			bb.get(content);
+			if (version == VERSION_SIGNED) {
```

Comment:
It's foreseeable that we'll have at least 2 further version increments in the short-to-medium term:

- New message id (version 30)
- Encryption support (version 31)

I'm a little worried about turning this class into a pile of spaghetti with a lot of `if (version == MY_COOL_NEW_VERSION)` branches everywhere. Is there an alternative approach?

How about some more object-oriented style? Version 29 is clearly an extension of 28. There could also be an abstract parent class for shared functionality and bootstrapping the instance creation with a static factory method `from(ByteBuffer)` in this parent class. A few pointers below, I'll leave it to you to play around with the idea if you like it:

```
public abstract class StreamrBinaryMessage {
...
    // There could be a shared version-aware entry point, with the assumption that first byte is version byte
    public static StreamrBinaryMessage from(ByteBuffer bb) {
        byte version = bb.get();
        if (version == 28) {
            return new StreamrBinaryBufferV28(bb);
        } else if (version == 29) {
            return new StreamrBinaryBufferV29(bb);
        } else {
            // throw
        }
    }
}
```
```
public class StreamrBinaryMessageV28 extends StreamrBinaryMessage {
    // Pretty much what the old StreamrBinaryMessage was
}
```
```
public class StreamrBinaryMessageV29 extends StreamrBinaryMessageV28 {
    // New fields only here
	private final byte signatureType;
	private final byte[] addressBytes;
	private final byte[] signatureBytes;

    // Examples of cool stuff the inheritance gives you
    public StreamrBinaryMessageV29(ByteBuffer bb) {
        super(bb);

        // Only read new fields here
        signatureType = bb.get();
	if (signatureType == SIGNATURE_TYPE_ETH) {
		addressBytes = new byte[20];
		bb.get(addressBytes);
		signatureBytes = new byte[65];
		bb.get(signatureBytes);
	} else if (signatureType == SIGNATURE_TYPE_NONE) {
		addressBytes = null;
		signatureBytes = null;
	} else {
		throw new IllegalArgumentException("Unknown signature type: "+signatureType);
	}
    }

    // Another nice perk
    @Override
    public int sizeInBytes() {
        return super.sizeInBytes() + (signatureType == SIGNATURE_TYPE_ETH ? 86 : 1);
    }
}
```